%% BioMed_Central_Tex_Template_v1.06
%%                                      %
%  bmc_article.tex            ver: 1.06 %
%                                       %

%%IMPORTANT: do not delete the first line of this template
%%It must be present to enable the BMC Submission system to
%%recognise this template!!

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                                                 %%
%% For instructions on how to fill out this Tex template:          %%
%% http://www.biomedcentral.com/info/authors/                      %%
%%                                                                 %%
%% Please do not use \input{...} to include other tex files.       %%
%%                                                                 %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%% additional documentclass options:
%  [doublespacing]
%  [linenumbers]   - put the line numbers on margins

%\documentclass[twocolumn]{bmcart}% uncomment this for twocolumn layout and comment line below
\documentclass{bmcart}

%%% Load packages
%\usepackage{amsthm,amsmath}
%\RequirePackage{natbib}
%\RequirePackage[authoryear]{natbib}% uncomment this for author-year bibliography
%\RequirePackage{hyperref}
\usepackage{soul}
\usepackage[utf8]{inputenc} %unicode support
%\usepackage[applemac]{inputenc} %applemac support if unicode package fails
%\usepackage[latin1]{inputenc} %UNIX support if unicode package fails


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                             %%
%%  If you wish to display your graphics for   %%
%%  your own use using includegraphic or       %%
%%  includegraphics, then comment out the      %%
%%  following two lines of code.               %%
%%  NB: These line *must* be included when     %%
%%  submitting to BMC.                         %%
%%  All figure files must be submitted as      %%
%%  separate graphics through the BMC          %%
%%  submission process, not included in the    %%
%%  submitted article.                         %%
%%                                             %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


\def\includegraphic{}
\def\includegraphics{}



%%% Put your definitions there:
\startlocaldefs
\newcommand{\matr}[1]{\mathbf{#1}}
\newcommand{\trans}[1]{{\cal #1}}
\newcommand{\eqref}[1]{Equation~\ref{#1}}

\newcommand{\statespace}{\Phi}
\newcommand{\state}{\phi}
\newcommand{\statevec}{\vec{p}}
\newcommand{\ratematrix}{\matr{R}}
\newcommand{\offdiagratematrix}{\matr{Q}}
\newcommand{\vectorofones}{\vec{1}}
\newcommand{\diagmatrix}[1]{\mbox{diag}(#1)}
\newcommand{\eqmvec}{\vec{\pi}}
\newcommand{\condmatrix}{\matr{M}}
\newcommand{\felsvec}{\vec{F}}
\newcommand{\unitvec}{\vec{\Delta}}
\newcommand{\pointprod}{\circ}
\newcommand{\scalarprod}{\cdot}
\newcommand{\initvec}{\vec{\rho}}
\endlocaldefs


%%% Begin ...
\begin{document}

%%% Start of article front matter
\begin{frontmatter}

\begin{fmbox}
\dochead{Commentary}

\title{Solving the Master Equation for Indels}

\author[
   addressref={aff1},                   % id's of addresses, e.g. {aff1,aff2}
   corref={aff1},                       % id of corresponding address, if any
   email={ihh@berkeley.edu}   % email address
]{\inits{IH}\fnm{Ian H} \snm{Holmes}}

\address[id=aff1]{%                           % unique id
  \orgname{Dept of Bioengineering, University of California}, % university, etc
  \postcode{94720}                                % post or zip code
  \city{Berkeley},                              % city
  \cny{USA}                                    % country
}

%\begin{artnotes}
%\note{Sample of title note}     % note to the article
%\note[id=n1]{Equal contributor} % note, connected to author
%\end{artnotes}

\end{fmbox}% comment this for two column layout

\begin{abstractbox}

\begin{abstract} % abstract
%\parttitle{Background} %if any
Despite long-heralded promise of putting sequence alignment on the same footing as statistical phylogenetics,
theorists have struggled to develop time-dependent evolutionary models for indels that are as tractable
as the analogous models for substitution events.

%\parttitle{} %if any
We discuss progress in this area,
in view of recent articles on the calculation of time-dependent gap length distributions in pairwise alignments
and current approaches for extending these approaches from ancestor-descendant pairs to phylogenetic trees.
\end{abstract}

\begin{keyword}
\kwd{phylogenetics}
\kwd{alignment}
\kwd{indels}
\end{keyword}

% MSC classifications codes, if any
%\begin{keyword}[class=AMS]
%\kwd[Primary ]{}
%\kwd{}
%\kwd[; secondary ]{}
%\end{keyword}

\end{abstractbox}
%
%\end{fmbox}% uncomment this for twcolumn layout

\end{frontmatter}

%%%%%%%%%%%%%%%%
%% Background %%
%%
\section*{Background}

Models of sequence evolution, formulated as continuous-time discrete-state Markov chains,
are central to statistical phylogenetics and bioinformatics.
As descriptions of the process of nucleotide or amino acid substitution,
their earliest uses were to estimate evolutionary distances \cite{JukesCantor69},
parameterize sequence alignment algorithms \cite{DayhoffEtal72},
and construct phylogenetic trees \cite{Felsenstein81}.
Variations on these models, including extra latent variables,
have been used to estimate spatial variation in evolutionary rates \cite{Yang93,Yang94};
these patterns of spatial variation have been used to
predict exon structures of protein-coding genes \cite{PedersenHein2003,SiepelHaussler04b},
foldback structure of non-coding RNA genes \cite{PedersenEtAl2006,PollardEtAl2006},
regulatory elements \cite{PedersenEtAl04},
ultra-conserved elements \cite{SiepelEtAl2005},
protein secondary structures \cite{GoldmanEtAl96},
and transmembrane structures \cite{LioGoldman99}.
They are widely used to reconstruct ancestral sequences \cite{BlanchetteEtAl2004,UgaldeEtAl2004,Liberles2007,OrtlundEtAl2007,GaucherEtAl2008,AshkenazyEtAl2012,AlcolombriEtAl2011,SantiagoOrtizEtAl2015,ZakasEtAl2016},
a method that is finding increasing application in synthetic biology \cite{Liberles2007,AlcolombriEtAl2011,SantiagoOrtizEtAl2015,ZakasEtAl2016}.
Trees built using substitution models used to classify species \cite{pmid26385966},
predict protein function \cite{EngelhardtEtAl2005},
inform conservation efforts \cite{pmid25561668},
or identify novel pathogens \cite{pmid12690091}.
In the analysis of rapidly evolving pathogens, these methods are used to
uncover population histories \cite{pmid15703244},
analyze transmission dynamics \cite{pmid22927414},
reconstruct key transmission events \cite{pmid27783600},
and predict future evolutionary trends \cite{pmid27774306}.
There are many other applications;
the ones listed above were selected to give some idea of how influential these models have been.

Continuous-time Markov chains describe evolution in a state space $\statespace$,
for example the set of nucleotides $\statespace = \{ \mbox{A,C,G,T} \}$.
The stochastic process $\state(t)$ at any given instant of time, $t$,
takes one of the values in $\statespace$.
We write $\statevec(t)$ for the vector describing the marginal probability distribution of the process
at a single point in time:
$\statevec_i(t) = P(\state(t) = i)$.
The time-evolution of this vector is governed by a {\em master equation}
\begin{equation}
\frac{d}{dt} \statevec(t) = \statevec(t) \ratematrix
\label{MasterEquation}
\end{equation}
where, for $i,j \in \statespace$ and $i \neq j$,
$\ratematrix_{ij}$ is the instantaneous rate of mutation from state $i$ to state $j$;
for probabilistic normalization of \eqref{MasterEquation},
we then require that
\begin{equation}
\ratematrix_{ii} = -\sum_{j \in \statespace, j \neq i} \ratematrix_{ij}
\label{DiagonalRates}
\end{equation}

The probability distribution of this process at equilibrium is given by the vector $\eqmvec$,
which must satisfy the equation
\[
\eqmvec \ratematrix = \vec{0}
\]

The general solution to \eqref{MasterEquation} can be
written $\statevec(t) = \statevec(0) \condmatrix(t)$
where $\condmatrix(t)$ is the {\em matrix exponential}
\begin{equation}
\condmatrix(t) = \exp ( \ratematrix t )
\label{MatrixExponential}
\end{equation}

Entry $\condmatrix_{ij}(t)$ of this matrix is the probability
$P(\state(t)=j|\state(0)=i)$
that, conditional on starting in state $i$,
the system will after time $t$ be in state $j$.
It follows, by definition, that this matrix satisfies the
Chapman-Kolmogorov forward equation:
\begin{equation}
\condmatrix(t) \condmatrix(u) = \condmatrix(t+u)
\label{ChapmanKolmogorov}
\end{equation}
That is, if $\condmatrix_{ij}(t)$ is the probability
that state $i$ will, after a finite time interval $t$, have evolved into state $j$,
and $\condmatrix_{jk}(u)$ is the analogous probability that state $j$ will after time $u$ evolve into state $k$,
then summing out $j$ has the expected result:
\[
\sum_{j \in \statespace} \condmatrix_{ij}(t) \condmatrix_{jk}(u) = \condmatrix_{ik}(t+u)
\quad \forall i,k \in \statespace
\]
This is one way of stating the defining property of a Markov chain:
its lack of historical memory previous to its current state.
\eqref{MasterEquation} is just an instantaneous version of this equation,
and \eqref{ChapmanKolmogorov} is the same equation in matrix form.

The conditional likelihood for an ancestor-descendant pair can be converted
into a phylogenetic likelihood for a set of extant taxon states $S$ related by a tree $T$,
as follows.
(We assume for convenience that $T$ is a binary tree, though relaxing this constraint is straightforward.)

We first compute, for every node $n$ in the tree,
the probability $\felsvec_i(n)$
of all observed states at leaf nodes descended from node $n$,
conditioned on node $n$ being in state $i$.
This is given by Felsenstein's {\em pruning recursion}:
\begin{equation}
\felsvec(n) = \left\{
\begin{array}{ll}
\displaystyle
\left( \condmatrix(t_{nl}) \felsvec(l) \right)
\pointprod
\left( \condmatrix(t_{nr}) \felsvec(r) \right)
& \mbox{if $n$ is an internal node with children $l,r$} \\
\displaystyle
\vec{\Delta}(s_n)
& \mbox{if $n$ is a leaf node in state $s_n$}
\end{array}
\right.
\label{Felsenstein}
\end{equation}
where $t_{mn}$ denotes the length of the branch from tree node $m$ to tree node $n$.
We have used the notation
$\unitvec(j)$ for the unit vector in dimension $j$,
and the symbol
$\pointprod$ to denote the {\em Hadamard product} (also known as the {\em pointwise product}),
defined such that for any two vectors $\vec{u},\vec{v}$ of the same size:
$(\vec{u} \pointprod \vec{v})_i = \vec{u}_i \vec{v}_i$.

Supposing that node $1$ is the root node of the tree,
and that the distribution of states at this root node is given by $\initvec$,
we can write the likelihood as
\begin{equation}
P(S|T,\ratematrix,\initvec) = \initvec \scalarprod \felsvec(1)
\label{Likelihood}
\end{equation}
where $\vec{u} \scalarprod \vec{v}$ denotes the scalar product of $\vec{u}$ and $\vec{v}$.
It is common to assume that the root node is at equilibrium,
so that $\initvec = \eqmvec$.

As mentioned above, this mathematical approach is fundamental to statistical phylogenetics
and many applications in bioinformatics.
For small state spaces $\statespace$, such as (for example) the 20 amino acids or 61 sense codons,
the matrix exponential $\condmatrix(t)$ in \eqref{MatrixExponential} can be solved exactly and practically
by the technique of spectral decomposition (i.e. finding eigenvalues and eigenvectors).
Such an approach informs the Dayhoff PAM matrix.
It was also solved for certain specific parametric forms of the rate matrix $\ratematrix$
by Jukes and Cantor \cite{JukesCantor69}, Kimura \cite{Kimura80}, Felsenstein \cite{Felsenstein81},
and Hasegawa {\em et al} \cite{HasegawaEtal85}, among others.
This approach is used by all likelihood-based phylogenetics tools, such as
RevBayes \cite{HohnaEtAl2016},
BEAST \cite{pmid17996036},
RAxML \cite{pmid16928733},
HyPhy \cite{pmid15509596},
PAML \cite{pmid17483113},
PHYLIP \cite{Felsenstein1989},
TREE-PUZZLE \cite{pmid11934758},
and XRate \cite{pmid22693624}.
Many more bioinformatics tools use the Dayhoff PAM matrix or other substitution matrix
based on an underlying master equation of the form \eqref{MasterEquation}.

\color{red}
\subsubsection*{Homogeneity, stationarity, and reversibility}

There exists a deep literature on Markov chains, to which this brief survey cannot remotely do justice,
but several concepts must be mentioned in order to survey progress in this area.

A Markov chain is {\em time-homogeneous} if the elements of the
rate matrix $\ratematrix$ in \eqref{MasterEquation} are themselves independent of time.
If a Markov chain is time-homogeneous and is known to be in equilibrium at a given time,
for example $\statevec(0) = \eqmvec$,
then (absent any other constraints) it will be in equilibrium at all times;
such a chain is referred to as being {\em stationary}.

A Markov chain is {\em reversible} if it satisfies the instantaneous {\em detailed balance} condition
$\eqmvec_i \ratematrix_{ij} = \eqmvec_j \ratematrix_{ji}$,
or its finite-time equivalent
$\eqmvec_i \condmatrix_{ij} = \eqmvec_j \condmatrix_{ji}$.
This amounts to a symmetry constraint on the parameter space of the chain
(specifically, the matrix with elements $S_{ij} = \sqrt{\eqmvec_i / \eqmvec_j} \ratematrix_{ij}$ is symmetric)
which has several convenient advantages: it effectively halves the number of parameters that must be
determined, it eases some of the matrix manipulations (symmetric matrices have real eigenvalues
and the algorithms to find them are generally more stable),
and it allows for some convenient manipulations, such as the so-called
{\em pulley principle} allowing for arbitrary re-rooting of the tree \cite{Felsenstein81}.
From another angle, however, these supposed advantages may be viewed as drawbacks:
reversibility is a simplification which ignores some unreversible aspects of real data,
limits the expressiveness of the model, and makes the root node placement statistically unidentifiable.

Stationarity has similar advantages and drawbacks.
If we assume the process was started at equilibrium, that is one less set of parameters to worry about
(since the equilibrium distribution is implied by the process itself), but it also renders the model
less expressive and makes some kinds of inference impossible.

The early literature on substitution models involved generalizing from rate matrices $\ratematrix$
characterized only by a single rate parameter \cite{JukesCantor69},
to symmetry-breaking versions that allowed for different transition and transversion rates
 \cite{Kimura80}, non-uniform equilibrium distributions over nucleotides \cite{Felsenstein81},
and combinations of the above \cite{HasegawaEtal85}.
These models are all, however, reversible.
A good deal of subsequent research has gone into the problem, in various guises, of generalizing results obtained
for reversible, homogeneous and/or stationary models to the analogous irreversible, nonhomogeneous and nonstationary models.
For examples, see \cite{pmid9600890,pmid22644340,pmid15210317}.
\color{black}

\subsubsection*{From individual residues to whole sequences}

The question naturally arises: how to extend the model to describe the evolution of an entire sequence,
not just individual sites?
As long as the allowed mutations are restricted to point substitutions
and their mutation rates are independent of flanking context,
then the extension to whole sequences is trivially easy:
one can simply multiply together probabilities of independent sites.

However, many kinds of mutation violate this assumption of site independence;
most notably context-dependent substitutions and indels, where the rates depend on neighboring sites.
For these mutations the natural approach is to extend the state space $\statespace$
to be the set of all possible sequences over a given alphabet
(for example, the set of all DNA or protein sequences).
This state space is (countably) infinite;
we can still discuss Equations~\ref{MasterEquation}-\ref{Felsenstein}
on an infinite state space,
but solution by brute-force enumeration of eigenvalues and eigenvectors is no longer feasible,
\color{red}
except in special cases where there is explicit structure to the rate matrix that allows identification of the eigensystem by algebraic approaches
\cite{LunterHein04,pmid21827770,pmid26135206}.
\color{black}

It has turned out that whole-sequence evolutionary models have proved quite challenging for theorists.
There is extensive evidence suggesting that indels, in particular, can be profoundly informative to phylogenetic studies,
and to applications of phylogenetics in sequence analysis \cite{pmid8445636,pmid15276848,pmid18578882,pmid23475937,pmid19958081,pmid16354754}.
The field of efforts to unify alignment and phylogeny,
and to build a theoretical framework for the evolutionary analysis of indels,
has been dubbed
{\em statistical alignment} by Hein, one of its pioneers \cite{HeinEtal2000}.

%%%%%%%%%%%%%%%%
%% Main text %%
%%
\section*{Discussion}

We focus only on ``local'' mutations: mostly indel events (which may include local duplications),
but also context-dependent substitutions.
This is not because ``nonlocal'' events (such as rearrangements) are unimportant,
but rather that they tend to defy phylogenetic reconstruction due to the rapid proliferation of possible histories
after even a few such events \cite{pmid9773350}.

The discussion here is separated into two parts.
In the first part, we discuss
the master equation (\eqref{MasterEquation})
and exact solutions thereof (Equations~\ref{MatrixExponential}),
along with various approximations and their departure from the Chapman-Kolmogorov ideal (\eqref{ChapmanKolmogorov}).
This is an area in which recent progress has been reported in this journal.
In the second part, we review
the extension from pairwise probability distributions
to phylogenetic likelihoods of multiple sequences,
using analogs of Felsenstein's pruning recursion (\eqref{Felsenstein}).

\subsection*{Solving the master equation}

We first begin with various approaches to finding the time-dependent probability distribution
of gap lengths in a pairwise alignment,
under several evolutionary models.

\color{red}
\subsubsection*{Exactly solved models on k-mer strings}

As an approach to models on strings of unbounded length, we can consider short motifs of $k$ residues.
These can still be considered as finite state-space models;
for example, a $k$-nucleotide model has $4^k$ possible states.
Several such models have been analyzed, including models on codons where $k=3$
\cite{pmid8281128,pmid7968486},
dinucleotides involved in RNA base-pairs where $k=2$ \cite{pmid7475089,KnudsenHein99,pmid17884102},
and models over sequences of arbitrary length $k$ \cite{LunterHein04,pmid26135206}.

Mostly, these models handle short sequences (motifs) and do not allow the sequence length to change over time
(so they model only substitutions and not indels).
Some of the later models do allow the sequence length to change via insertions or deletions
\cite{pmid26135206} though these models have not yet been analyzed in a way that would allow the
computation of alignment likelihoods for sequences of realistic lengths.
\color{black}

\subsubsection*{Exactly solved models on strings of unbounded length}

It is a remarkable reflection on the extremely challenging nature of this problem
that, to date, the only exactly solved indel model on strings
is the TKF91 model, named after the authors' initials and date of publication
of this seminal paper \cite{ThorneEtal91}.
While there has been progress in developing approximate models in the 25 years since the publication of this paper,
and in extending it from pairwise to multiple sequence alignment,
it remains the only model for which
\begin{enumerate}
\item the state space $\statespace$ is the set of all sequences (strings) over a finite alphabet,
\item the state space is ergodically explored by substitutions and indels
(so there is a valid alignment and evolutionary trajectory between any two sequences $\state(0)$ and $\state(t)$),
\item \eqref{MatrixExponential} can be calculated exactly
 (specifically, as a sum over alignments,
  where the individual alignment likelihoods can be written in closed form).
\end{enumerate}

The TKF91 model allows single-residue context-independent events only.
These include (i) single-residue substitutions,
(ii) single-residue insertions (with the inserted residue drawn from the equilibrium distribution
of the substitution process), and
(iii) single-residue deletions (whose rates are independent of the residue being deleted).
The rates of all these mutation events are independent of the flanking sequence.

This process is equivalent to a linear birth-death model with constant immigration
\cite{Feller71}.
Thorne {\em et al} showed that 
an ancestral sequence can be split into independently evolving zones, one for each ancestral residue
(or ``links'', as they call them).
This leads to the very appealing result that the length distribution for observed gaps is geometric,
which conveniently allows the joint probability $P(\state(0),\state(t))$ to be expressed
as a paired-sequence Hidden Markov Model or ``Pair HMM'' \cite{HolmesBruno2001}.
The conditional probability $P(\state(t)|\state(0))$
can similarly be expressed as a weighted finite-state transducer \cite{Holmes2003,WestessonEtAl2012,BouchardCote2013}.
\color{red}
Some interesting discussion of why the TKF91 model should be solvable at all can be found in
\cite{Metzler2003} and in \cite{pmid22644340}.
\color{black}

There are several variations on the TKF91 model.
The case where there are no indels at all, only substitutions, can be viewed as a special case of TKF91,
and can of course be solved exactly, as is well known.
Another variation on the TKF91 model constrains the total indel rate to be independent of sequence
length \cite{BouchardCoteJordan2013}.

In the following section we cover some variants that use different state spaces.

\subsubsection*{Exactly solved models on state spaces other than strings}

It is difficult to extend TKF91 to more realistic models wherein indels (or substitutions)
can affect multiple residues at once.
In such models, the fate of adjacent residues is no longer independent, since a single event can span multiple sites.

As a way around this difficulty, several researchers have developed
evolutionary models where the state is not a pure DNA or protein sequence,
but includes some extra ``hidden'' information,
such as boundaries, markers or other latent structure.
In some of these models the sequence of residues is replaced by a sequence of indivisible fragments,
each of which can contain more than one residue \cite{ThorneEtal92,Metzler2003,RivasEddy2015}.
These includes the TKF92 model \cite{ThorneEtal92} which is, essentially, TKF91 with residues replaced by
fragments (so the alphabet itself is the countably infinite set of all sequences over some other, finite alphabet).
Other models approximate indels as a kind of substitution that temporarily hides a residue,
by augmenting the DNA or protein alphabet with an additional gap character \cite{McGuireEtAl2001,Rivas05,RivasEddy2008}.

These models can be used to calculate some form of likelihood for a pairwise alignment of two sequences,
but since this likelihood is not derived from an underlying instantaneous model of indels,
the equations do not, in general, satisfy the Chapman-Kolmogorov forward equation
(\ref{ChapmanKolmogorov}).
That is, the probability of evolving from $i$ to $k$ comes out differently
depending on whether or not one conditions on an intermediate sequence $j$.
Clearly, something about this ``seems wrong'':
the failure to obey \eqref{ChapmanKolmogorov} illustrates the {\em ad hoc}
nature of these approaches.
Ezawa \cite{Ezawa2016b} describes the Chapman-Kolmogorov property
(\eqref{ChapmanKolmogorov}) as {\em evolutionary consistency};
it can also be regarded as being the defining property
of any correct solution to a continuous-time Markov chain.
The above-mentioned approaches may be evolutionarily consistent if the state space
is allowed to include the extra information that is introduced to make the model tractable,
such as fragment boundaries.

Models which allow for heterogeneity of indel and substitution rates along the sequence
fall into this category. The usual way of allowing for such spatial variation
in substitution models is to assume a latent parameterization associated with each site \cite{Yang93,Yang94}.
For indel models, this latent information must be extended to include hidden site boundaries \cite{RivasEddy2015}.

\subsubsection*{Exactly solved models on graphs}

Another variation on TKF91 is the TKF Structure Tree, which describes the evolutionary behavior
of RNA structures with stem and loop regions which are subject to insertion and deletion \cite{Holmes2004}.
Rather than describing the evolution of a sequence,
this model essentially captures the time-evolution grammar of a tree-like graph whose
individual edges are evolving according to the TKF91 model.
Other evolutionary models have made use of graph grammars,
for example to model pseudoknots \cite{MatsuiEtAl2005} or context-dependent indels \cite{HickeyBlanchette2011}.

\subsubsection*{Finite-event trajectories}

In tackling indel models where the indel events can insert or delete multiple residues at once,
several authors have used the approximation that indels never overlap,
so that any observed gap corresponds to a single indel event.
This approximation, which is justified if one is considering evolutionary timespans $t \ll 1/\delta$
where $\delta$ is the indel rate per site,
considerably simplifies the task of calculating gap probabilities
\cite{KnudsenMiyamoto2003,RedelingsSuchard2005,SuchardRedelings2006,RedelingsSuchard2007,WestessonEtAlArxiv2011,WestessonEtAl2012,WestessonBarquistHolmes2012}.

At longer timescales, it is necessary to consider multiple-event trajectories,
but (as a simplifying approximation) one can still truncate the trajectory at a finite number of events.
A problem with this approach is that many different trajectories will generally be consistent with an observed mutation.
Summing over all such trajectories, to compute the probability of observing a particular configuration after finite time
(e.g. the observed gap length distribution),
is a nontrivial problem.

In analyzing the {\em long indel} model,
a generalization of TKF91 with arbitrary length distributions for instantaneous indel events,
Mikl\'{o}s {\em et al} \cite{MiklosLunterHolmes2004}
make the claim that the existence of a conserved residue implies the alignment probability is factorable at that point
 (since no indel has ever crossed the boundary).
They use a numerical sum over indel trajectories to approximate the probability distribution of observed gap lengths.
Although they used a reversible model, their approach generalizes readily to irreversible models.
This work builds on an earlier model which allows long insertions, but only single-residue deletions \cite{MiklosEtal2001}.
Recent work by Ezawa has put this finite-event approximation on a more solid footing
by developing a rigorous algebraic definition of equivalence classes for event trajectories \cite{Ezawa2016a,Ezawa2016b,Ezawa2016bErratum}.

Solutions obtained using finite-event approximations will not exactly satisfy \eqref{ChapmanKolmogorov}.
There will be some error in the probability, and in general the error will be greater
on longer branches, as the main assumption behind the approximation
(that there are no overlapping indels in the time interval, or that there is a finite
limit to the number of overlapping indels)
starts to break down.
However, since these are principled approximations, it should be possible to form some conclusions
as to the severity of the error, and its dependence on model parameters.
Simulation studies have also been of some help in assessing the error of these approximations.

\subsubsection*{Taylor series approximations}

For context-dependent substitution processes,
such as models that include methylation-induced CpG-deamination,
a clever approach was developed in \cite{LunterHein04}.
Rather than considering a finite-event trajectory, they develop an explicit Taylor series for the matrix exponential
(\eqref{MatrixExponential}) and then truncate this Taylor series.
\color{red}
Specifically, the rate matrix for a finite-length sequence is constructed
as a sum of rate matrices operating locally on the sequence, using the Kronecker sum $\oplus$
and Kronecker product $\otimes$ to concatenate rate matrices.
These operators may be understood as follows, for an alphabet $\Omega$ of $N$ symbols:
suppose that ${\cal K}_m$ is the set of all matrices indexed by $m$-mers,
so that if $\matr{A} \in {\cal K}_m$, then $\matr{A}$ is an $N^m \times N^m$ matrix.
Let $i,j$ be $m$-mers, $k,l$ be $n$-mers, and $ik, jl$ the concatenated $m+n$-mers.
If $\matr{A} \in {\cal K}_m$ and $\matr{B} \in {\cal K}_n$
then $\matr{A} \oplus \matr{B}$ and $\matr{A} \otimes \matr{B}$ are both in ${\cal K}_{m+n}$
and are specified by
\begin{eqnarray}
  \left( \matr{A} \otimes \matr{B} \right)_{ik,jl} & = & \matr{A}_{i,j} \matr{B}_{k,l} \\
  \left( \matr{A} \oplus \matr{B} \right)_{ik,jl} & = & \delta(k=l) \matr{A}_{i,j} + \delta(i=j) \matr{B}_{k,l}
\end{eqnarray}
Furthermore, suppose $\matr{O}_n \in {\cal K}_n$ is the $N^k \times N^k$ null matrix containing only zeroes.
Then $\matr{O}_m \oplus \matr{B}$ commutes with $\matr{A} \oplus \matr{O}_n$,
and $\exp(\matr{A} \oplus \matr{B}) = \exp(\matr{A}) \otimes \exp(\matr{B})$.
The rate matrix $\ratematrix$ for a length-$L$ sequence operated on locally by a context-sensitive rate matrix $\matr{A} \in {\cal K}_m$ can be written as a sum of the form
\[
\ratematrix = \sum_{n=0}^{L-m} \matr{O}_n \oplus \matr{A} \oplus \matr{O}_{L-m-n}
\]
Commuting terms in the Taylor series for $\exp(\ratematrix t)$ can be systematically rearranged into an explicit dynamic programming recursion.
This approach was first used by \cite{LunterHein04} and further developed, with more in-depth discussion of the associated eigensystem, by \cite{pmid21827770,pmid26135206}.
\color{black}
It remains to be seen if such an approach can be used for indel models.

\subsubsection*{Simulation studies}

Such is the difficulty of solving long indel models that several authors have performed simulations
to investigate the empirical gap length distributions that are observed after finite time intervals
for various given instantaneous indel-rate models.
These observed gaps can arise from multiple overlapping indel events, in ways that have so far defied
straightforward algebraic characterization.

In recent work, Rivas and Eddy \cite{RivasEddy2015}
have shown that if an underlying model has a simple geometric length distribution over instantaneous indel events,
the observed gap length distribution at finite times cannot be geometric.
This result suggests that HMM-like models,
which are most efficient at modeling geometric length distributions
(or related distributions for sums of geometric random variates, such as negative binomial distributions),
may be fundamentally limited in their ability to fully describe indels.
However, generalized HMMs, which can model arbitrary length distributions at the cost of some
computational efficiency \cite{BurgeKarlin97}, may be up to the task;
for example, the abovementioned ``long indel'' model of Mikl\'{o}s {\em et al} uses a generalized Pair HMM \cite{MiklosLunterHolmes2004}.
Rivas and Eddy report simulation studies supporting this result,
and go on to propose several models incorporating hidden information
(such as fragment boundaries, {\em a la} TKF92)
which have the advantage of being good fits to HMMs for their finite-time distributions.

The recent work of Ezawa has some parallels, but also differences \cite{Ezawa2016a,Ezawa2016b,Ezawa2016bErratum}.
Ezawa criticizes over-reliance on HMM-like models, and insists on a systematic derivation from simple instantaneous models.
He puts the intuition of Mikl\'{o}s et al \cite{MiklosLunterHolmes2004}
on a more formal footing by introducing an explicit notation for indel trajectories
and the concept of ``local history set equivalence classes'' for equivalent trajectories.
Ezawa uses this concept to prove that alignment likelihoods for long-indel and related models are indeed factorable,
and investigates by simulation and analysis the relative contribution
of multiple-event trajectories to gap length distributions.


\subsection*{Extending from pairs to trees}

We now move to the extension of results on pairwise alignments,
such as TKF91 and the ``long indel'' model, to multiple alignments (supported by phylogenies).
In considering this extension,
we can divide the approaches into two classes:
those that use Markov Chain Monte Carlo (MCMC), and those that do not.

\subsubsection*{Approaches based on MCMC sampling}

MCMC is the most principled approach to integrating phylogeny with multiple alignment.
In principle an MCMC algorithm for phylogenetic alignment can yield the posterior distribution of
alignments, trees, and parameters for any model whose pairwise distribution can be computed.
This includes long indel models and also, in principle, other effects such as context-dependent substitutions.

Of the MCMC methods reported in the literature,
some just focus on alignment and ancestral sequence reconstruction \cite{HolmesBruno2001};
others on simultaneous alignment and phylogenetic reconstruction
\cite{RedelingsSuchard2005,SuchardRedelings2006,RedelingsSuchard2007,NovakEtAl2008,BouchardCoteEtAl2009,WestessonBarquistHolmes2012};
some also include estimation of evolutionary parameters such as $dN/dS$ \cite{Redelings2014};
and some (focused on RNA sequences) attempt prediction of secondary structure \cite{ArunapuramEtAl2013,MeyerMiklos2007}.

In practise these all use HMMs, or dynamic programming of some form,
in common with the methods of the following section.

\subsubsection*{Approaches based on automata theory}

The dynamic programming recursion for pairwise alignment reported for the TKF91 model \cite{ThorneEtal91}
can be exactly extended to alignment of multiple sequences given a tree \cite{Hein2001,LunterSongMiklosHein2003}.
This works essentially because the TKF91 joint distribution over ancestor and descendant sequences
can be represented as a Pair HMM;
the multiple-sequence version is a multi-sequence HMM \cite{HolmesBruno2001}.

This approach can be generalized, using {\em finite-state transducer} theory.
Transducers were originally developed as modular representations of sequence-transforming operations
for use in speech recognition \cite{MohriPereiraRiley2000}.
In bioinformatics, they offer (among other things)
a systematic way of extending HMM-like pairwise alignment likelihoods to trees
\cite{pmid7584457,WestessonEtAl2012,BouchardCote2013,IndelHistorian}.
Other applications of transducer models in bioinformatics have included
copy number variation in tumors \cite{pmid24743184},
protein family classification \cite{pmid12804091},
DNA-protein alignment \cite{pmid15123596}
and error-correcting codes for DNA storage \cite{DBLP:journals/corr/Holmes16}.

A finite-state transducer is a state machine that reads an input tape and writes to an output tape \cite{Mealy55}.
A probabilistically weighted finite-state transducer is the same, but its behavior is stochastic \cite{MohriPereiraRiley2000}.
For the purposes of bioinformatics sequence analysis,
a transducer can be thought of as being just like a Pair HMM;
except where a Pair HMM's transition and emission probabilities
have been normalized so as to describe joint probabilities,
a transducer's probabilities are normalized so as to describe conditional probabilities
like the entries of matrix $\condmatrix(t)$ (\eqref{MatrixExponential}).
More specifically, if $i$ and $j$ are sequences, then we can define the matrix entry $A_{ij}$
to be the Forward score for those two sequences using transducer $\trans{A}$.
Thus, the transducer is a compact encoding for a square matrix of countably infinite rank,
indexed by sequence states (rather than nucleotide or amino acid states).

The utility of transducers is that for many purposes they can be manipulated just like matrices.
If $\trans{A}$ and $\trans{B}$ are transducers representing matrices $\matr{A}$ and $\matr{B}$,
then there is a well-defined operation called {\em transducer composition}
yielding a transducer $\trans{A}\trans{B}$ that represents the matrix $\matr{A}\matr{B}$.
There are other well-defined transducer operations corresponding to the various other linear algebra
operations used in this paper:
the Hadamard product ($\pointprod$) corresponds to {\em transducer intersection},
the Kronecker product ($\otimes$) to {\em transducer concatenation},
and the scalar product ($\scalarprod$) and the unit vector ($\unitvec$) can also readily be constructed using transducers.
Consequently, \eqref{Felsenstein} can be interpreted directly
in terms of transducers \cite{WestessonEtAlArxiv2011,WestessonEtAl2012,BouchardCote2013}.

This has several benefits.
One is theoretical unification: \eqref{Felsenstein}, using the above
linear algebra interpretation of transducer manipulations,
turns out to be very similar to Sankoff's algorithm for phylogenetic multiple alignment \cite{Sankoff85}.
Thus is a famous algorithm in bioinformatics unified with a famous algorithm in likelihood phylogenetics
by using a tool from computational linguistics.
(This excludes the RNA structure-prediction component of Sankoff's algorithm;
that can, however, be included by extending the transducer
framework to pushdown automata \cite{BradleyHolmes2009}.)
Practically, the phylogenetic transducer can be used for alignment \cite{RedelingsSuchard2005,RedelingsSuchard2007},
parameter estimation \cite{Redelings2014}, and ancestral reconstruction \cite{WestessonEtAl2012},
with promising results for improved accuracy in multiple sequence alignment \cite{IndelHistorian}.

More broadly, we can think of the transducer as being in a family of methods
that combine phylogenetic trees (modeling the temporal structure of evolution)
with automata theory, grammars, and dynamic programming on sequences (modeling the spatial structure of evolution).
The TKF Structure Tree, mentioned above, is in this family too:
it can be viewed as a context-free grammar, or as a transducer with a pushdown stack \cite{Holmes2004}.

The HMM-like nature of TKF91, and the ubiquity of HMMs and dynamic programming in sequence analysis,
has motivated numerous approaches to indel analysis based on Pair HMMs
\cite{ThorneEtal92,KnudsenMiyamoto2003,WangKeightleyJohnson2006,RivasEddy2008,RivasEddy2015},
as well as many other applications of 
phylogenetic HMMs \cite{GoldmanEtAl96,PedersenHein2003,SiepelHaussler04,SiepelHaussler04b}
and phylogenetic grammars \cite{KnudsenHein99,KnudsenHein2003,PedersenEtAl04,PedersenEtAl2006,KlostermanEtAl2006,pmid22693624}.
In most of these models, an alignment is assumed fixed and the HMM or grammar used to partition it;
however, in principle, one can combine the ability of HMMs/grammars to model indels (and thus impute alignments)
with the ability to partition sequences into differently evolving regions.

%%%%%%%%%%%%%%%%
%% Conclusions %%
%%
\section*{Conclusions}

The promise of using continuous-time Markov chains to model indels
has been partially realized by automata-theoretic approaches
based on transducers and HMMs.
Recent work by Rivas and Eddy \cite{RivasEddy2015}
and by Ezawa \cite{Ezawa2016a,Ezawa2016b,Ezawa2016bErratum}
may be interpreted as both good and bad news for automata-theoretic approaches.

It appears that closed-form solutions for observed gap length distributions at finite times,
and in particular the geometric distributions that simple automata are good at modeling,
are still out of reach for realistic indel models,
and indeed (for simple models) have been proven impossible \cite{RivasEddy2015}.
Further, simulation results have demonstrated that geometric distributions are not
a good fit to the observed gap length distributions when the underlying indel model
has geometrically-distributed lengths for its instantaneous indel events \cite{RivasEddy2015,Ezawa2016a}.

That is the bad news (at least for automata).
The good news is that the simulation results also suggest that,
for short branches, the error may not be too bad to live with.
Approximate-fit approaches that are common in Pair HMM modeling and pairwise sequence alignment - such as using a mixture
of geometric distributions to approximate a gap length distribution (yielding a longer tail than can be modeled
using a pure geometric distribution) - may help bridge the accuracy gap \cite{DoEtAl2005}.
Given the power of automata-theoretic approaches, the best way forward
(in the absence of a closed-form solution) may be to embrace such approximations and live with the ensuing error.

Interestingly, the authors of the two recent simulation studies that prompted this commentary
come to different conclusions about the viability of automata-based dynamic programming approaches.
Ezawa \cite{Ezawa2016a,Ezawa2016b}, arguing that realism is paramount, advocates deeper study of the
gap length distributions obtained from simple instantaneous models - while acknowledging that such gap length distributions
may be more difficult to use in practice than the simple geometric distributions offered by HMM-like models.
Rivas and Eddy \cite{RivasEddy2015}, clearly targeting applications (particularly those such as profile HMMs),
work backward from HMM-like models toward evolutionary models with embedded
hidden information.
These models may be somewhat mathematically contrived, and thus (from a certain point of view) less realistic,
but they capture more empirically observed phenomena.

Whichever approach is used, these results are unambiguously good news for the theoretical study of indel processes.
The potential benefits of modeling alignment as an aspect of statistical phylogenetics are significant.
One can reasonably hope that the advance of theoretical work in this area 
will continue to inform advances in both bioinformatics and statistical phylogenetics.
After all, and in spite of the Cambrian explosion in bioinformatics sub-disciplines,
sequence alignment and phylogeny truly are closely related aspects of mathematical biology.



%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Backmatter begins here                   %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{backmatter}

\section*{Competing interests}
  The author declares that they have no competing interests.

\section*{Author's contributions}
    IH wrote the article.

\section*{Acknowledgements}
  The author thanks Kiyoshi Ezawa, Elena Rivas, Sean Eddy, Jeff Thorne,
  Benjamin Redelings, and Marc Suchard
  for productive conversations that have informed the writing of this piece.
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                  The Bibliography                       %%
%%                                                         %%
%%  Bmc_mathpys.bst  will be used to                       %%
%%  create a .BBL file for submission.                     %%
%%  After submission of the .TEX file,                     %%
%%  you will be prompted to submit your .BBL file.         %%
%%                                                         %%
%%                                                         %%
%%  Note that the displayed Bibliography will not          %%
%%  necessarily be rendered by Latex exactly as specified  %%
%%  in the online Instructions for Authors.                %%
%%                                                         %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% if your bibliography is in bibtex format, use those commands:
\bibliographystyle{bmc-mathphys} % Style BST file (bmc-mathphys, vancouver, spbasic).
\bibliography{references}      % Bibliography file (usually '*.bib' )
% for author-year bibliography (bmc-mathphys or spbasic)
% a) write to bib file (bmc-mathphys only)
% @settings{label, options="nameyear"}
% b) uncomment next line
%\nocite{label}

% or include bibliography directly:
% \begin{thebibliography}
% \bibitem{b1}
% \end{thebibliography}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                               %%
%% Figures                       %%
%%                               %%
%% NB: this is for captions and  %%
%% Titles. All graphics must be  %%
%% submitted separately and NOT  %%
%% included in the Tex document  %%
%%                               %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%
%% Do not use \listoffigures as most will included as separate files

%\section*{Figures}
%  \begin{figure}[h!]
%  \caption{\csentence{Sample figure title.}
%      A short description of the figure content
%      should go here.}
%      \end{figure}
%
%\begin{figure}[h!]
%  \caption{\csentence{Sample figure title.}
%      Figure legend text.}
%      \end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                               %%
%% Tables                        %%
%%                               %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%% Use of \listoftables is discouraged.
%%
%\section*{Tables}
%\begin{table}[h!]
%\caption{Sample table title. This is where the description of the table should go.}
%      \begin{tabular}{cccc}
%        \hline
%           & B1  &B2   & B3\\ \hline
%        A1 & 0.1 & 0.2 & 0.3\\
%        A2 & ... & ..  & .\\
%        A3 & ..  & .   & .\\ \hline
%      \end{tabular}
%\end{table}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                               %%
%% Additional Files              %%
%%                               %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%\section*{Additional Files}
%  \subsection*{Additional file 1 --- Sample additional file title}
%    Additional file descriptions text (including details of how to
%    view the file, if it is in a non-standard format or the file extension).  This might
%    refer to a multi-page table or a figure.
%
%  \subsection*{Additional file 2 --- Sample additional file title}
%    Additional file descriptions text.


\end{backmatter}
\end{document}
