%% BioMed_Central_Tex_Template_v1.06
%%                                      %
%  bmc_article.tex            ver: 1.06 %
%                                       %

%%IMPORTANT: do not delete the first line of this template
%%It must be present to enable the BMC Submission system to
%%recognise this template!!

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                                                 %%
%% For instructions on how to fill out this Tex template:          %%
%% http://www.biomedcentral.com/info/authors/                      %%
%%                                                                 %%
%% Please do not use \input{...} to include other tex files.       %%
%%                                                                 %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%% additional documentclass options:
%  [doublespacing]
%  [linenumbers]   - put the line numbers on margins

%\documentclass[twocolumn]{bmcart}% uncomment this for twocolumn layout and comment line below
\documentclass{bmcart}

%%% Load packages
%\usepackage{amsthm,amsmath}
%\RequirePackage{natbib}
%\RequirePackage[authoryear]{natbib}% uncomment this for author-year bibliography
%\RequirePackage{hyperref}
\usepackage[utf8]{inputenc} %unicode support
%\usepackage[applemac]{inputenc} %applemac support if unicode package fails
%\usepackage[latin1]{inputenc} %UNIX support if unicode package fails


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                             %%
%%  If you wish to display your graphics for   %%
%%  your own use using includegraphic or       %%
%%  includegraphics, then comment out the      %%
%%  following two lines of code.               %%
%%  NB: These line *must* be included when     %%
%%  submitting to BMC.                         %%
%%  All figure files must be submitted as      %%
%%  separate graphics through the BMC          %%
%%  submission process, not included in the    %%
%%  submitted article.                         %%
%%                                             %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


\def\includegraphic{}
\def\includegraphics{}



%%% Put your definitions there:
\startlocaldefs
\newcommand{\matr}[1]{\mathbf{#1}}
\newcommand{\trans}[1]{{\cal #1}}
\newcommand{\eqref}[1]{Equation~\ref{#1}}
\endlocaldefs


%%% Begin ...
\begin{document}

%%% Start of article front matter
\begin{frontmatter}

\begin{fmbox}
\dochead{Research}

\title{Towards Computable Models of Sequence Evolution}

\author[
   addressref={aff1},                   % id's of addresses, e.g. {aff1,aff2}
   corref={aff1},                       % id of corresponding address, if any
   email={ihh@berkeley.edu}   % email address
]{\inits{IH}\fnm{Ian H} \snm{Holmes}}

\address[id=aff1]{%                           % unique id
  \orgname{Dept of Bioengineering, University of California}, % university, etc
  \postcode{94720}                                % post or zip code
  \city{Berkeley},                              % city
  \cny{USA}                                    % country
}

%\begin{artnotes}
%\note{Sample of title note}     % note to the article
%\note[id=n1]{Equal contributor} % note, connected to author
%\end{artnotes}

\end{fmbox}% comment this for two column layout

\begin{abstractbox}

\begin{abstract} % abstract
%\parttitle{Background} %if any
Despite long-recognized benefits to putting sequence alignment on the same footing as statistical phylogenetics,
theorists have struggled to develop time-dependent evolutionary models for indels that are as tractable
as the analogous models for substitution events.

%\parttitle{} %if any
We review progress in this area, with reference to
recent articles on the calculation of gap length distributions.
\end{abstract}

\begin{keyword}
\kwd{phylogenetics}
\kwd{alignment}
\kwd{indels}
\end{keyword}

% MSC classifications codes, if any
%\begin{keyword}[class=AMS]
%\kwd[Primary ]{}
%\kwd{}
%\kwd[; secondary ]{}
%\end{keyword}

\end{abstractbox}
%
%\end{fmbox}% uncomment this for twcolumn layout

\end{frontmatter}

%%%%%%%%%%%%%%%%
%% Background %%
%%
\section*{Background}

Probabilistic models of substitution are centrally important in statistical phylogenetics and bioinformatics

Used to estimate evolutionary time separating pairs of species,
construct phylogenetic trees,
parameterize sequence alignment algorithms.
Estimate spatial variation in evolutionary rates;
patterns of spatial variation have been used to
predict exon structures of protein-coding genes
and foldback structure of non-coding RNA genes,
as well as secondary structure of proteins.
Reconstruct ancestral sequences for synthetic biology.
Trees are used to classify species,
identify novel pathogens,
reconstruct population histories in viral epidemics,
identify transmissione events,
inform conservation efforts,
among many other applications.

At core, these methods are based on the continuous-time Markov chain.
We assume there is a state space $\Phi$,
for example the set of nucleotides $\Phi = \{ \mbox{A,C,G,T} \}$.
There is a stochastic process $\phi(t)$ that at any given instant of time, $t$,
takes one of the values in $\Phi$.
We write $\vec{p}(t)$ for the vector of probabilities
$\vec{p}_i(t) = P(\phi(t) = i)$.
The time-evolution of this vector is governed by a {\em master equation}
\begin{equation}
\frac{d}{dt} \vec{p}(t) = \vec{p}(t) \matr{R}
\label{MasterEquation}
\end{equation}
where, for $i,j \in \Phi$ and $i \neq j$,
$\matr{R}_{ij}$ is the instantaneous rate of mutation from state $i$ to state $j$;
for probabilistic normalization of \eqref{MasterEquation},
we then require that
\[
\matr{R}_{ii} = -\sum_{j \in \Phi, j \neq i} \matr{R}_{ij}
\]

The probability distribution of this process at equilibrium is given by the vector $\vec{\pi}$,
which must satisfy the equation
\[
\vec{\pi} \matr{R} = \vec{0}
\]

The general solution to \eqref{MasterEquation} can be expressed in terms of the matrix exponential
\begin{eqnarray}
\vec{p}(t) & = & \vec{p}(0) \matr{M}(t) \\
\matr{M}(t) & = & \exp ( \matr{R}t )
\label{MatrixExponential}
\end{eqnarray}

The entries of the matrix \eqref{MatrixExponential} give the conditional probability
that an ancestral state $i$ will, after time $t$, have evolved into a descendant state $j$.
If we multiply each row of $\matr{M}(t)$ by the prior distribution of the ancestral state,
then we obtain the time-varying matrix $\matr{Q}(t)$ describing the joint probability of ancestor and descendant:
\begin{eqnarray}
\matr{M}_{ij}(t) & = & P(\phi(t)=j|\phi(0)=i) \\
\matr{Q}_{ij}(t) & = & P(\phi(0)=i) \times \matr{M}_{ij}(t) \label{JointLikelihood} \\
& = & P(\phi(0)=i,\phi(t)=j)
\end{eqnarray}

It is worth noting that, by definition, matrix $\matr{M}(t)$ satisfies the
Chapman-Kolmogorov forward equation:
\begin{equation}
\matr{M}(t) \matr{M}(u) = \matr{M}(t+u)
\label{ChapmanKolmogorov}
\end{equation}
That is, if $M_{ij}(t)$ is the probability
that sequence $i$ will after time $t$ evolve into sequence $j$,
and $M_{jk}(u)$ is the probability that sequence $j$ will after time $u$ evolve into sequence $k$,
then summing out sequence $j$ has the expected result:
\[
M_{ij}(t) M_{jk}(u) = M_{ik}(t+u)
\]

The conditional likelihood for an ancestor-descendant pair can be converted
into a phylogenetic likelihood for a set of extant taxon states $S$ related by a tree $T$,
as follows.
(We assume for convenience that $T$ is a binary tree, though relaxing this constraint is straightforward.)

We first compute, for every node $n$ in the tree,
the probability $\vec{F}_i(n)$
of all observed states at leaf nodes descended from node $n$,
conditioned on node $n$ being in state $i$.
This is given by Felsenstein's pruning recursion:
\begin{equation}
\vec{F}(n) = \left\{
\begin{array}{ll}
\displaystyle
\left( \matr{M}(t_{nl}) \vec{F}(l) \right)
\otimes
\left( \matr{M}(t_{nr}) \vec{F}(r) \right)
& \mbox{if $n$ is an internal node with children $l,r$} \\
\displaystyle
\vec{\Delta}(s_n)
& \mbox{if $n$ is a leaf node with taxon state $s_n$}
\end{array}
\right.
\label{Felsenstein}
\end{equation}
where $t_{mn}$ denotes the length of the branch from tree node $m$ to tree node $n$.
We have used the notation
$\vec{\Delta}(j)$ for the unit vector in dimension $j$,
and the symbol
$\otimes$ to denote the {\em Hadamard product} (also known as the {\em pointwise product}),
defined such that for any two vectors $\vec{u},\vec{v}$ of the same size:
$(\vec{u} \otimes \vec{v})_i = \vec{u}_i \vec{v}_i$.

Supposing that node $1$ is the root node of the tree,
and that the distribution of states at this root node is given by $\vec{\rho}$,
we can write the likelihood as
\begin{equation}
P(S|T,\matr{R},\vec{\rho}) = \vec{\rho} \cdot \vec{F}(1)
\label{Likelihood}
\end{equation}
where $\vec{u} \cdot \vec{v}$ denotes the scalar product of $\vec{u}$ and $\vec{v}$.
It is common to assume that the root node is at equilibrium,
so that $\vec{\rho} = \vec{\pi}$.

As mentioned above, this mathematical approach is fundamental to statistical phylogenetics
and many applications in bioinformatics.
For small state spaces $\Phi$, \eqref{MatrixExponential} can be solved exactly
by the technique of spectral decomposition (i.e. finding eigenvalues and eigenvectors).
Such an approach informs the Dayhoff PAM matrix.
It was also solved for certain specific parametric forms of $\matr{R}$
by Jukes and Cantor in 1969, Kimura in 1980, Felsenstein in 1981, and Hasegawa {\em et al} in 1985.
This approach is used by all likelihood-based phylogenetics tools including
RevBayes, BEAST, RAxML, HyPhy, PAML, PHYLIP, TREE-PUZZLE, XRate.
Many more bioinformatics tools use the Dayhoff PAM matrix or other substitution matrix
based on an underlying master equation of the form \eqref{MasterEquation}.

Question naturally arises: how to extend the model to describe the evolution of an entire sequence,
not just individual sites?
As long as the allowed mutations are restricted to point substitutions
and their mutation rates are independent of flanking context,
then extension to whole sequences is trivially easy:
can simply multiply together probabilities of independent sites.

However, some mutations violate this assumption of site independence;
most notably context-dependent substitutions and indels.
For these mutations the obvious approach is to extend the state space $\Phi$
to be the set of all possible sequences over a given alphabet
(for example, the set of all DNA or protein sequences).
This state space is (countably) infinite;
we can still discuss Equations~\ref{MasterEquation}-\ref{Likelihood}
on an infinite state space,
but solution by brute-force enumeration of eigenvalues and eigenvectors is no longer possible.
It is for this reason that whole-sequence models have proved quite challenging for theorists.

%%%%%%%%%%%%%%%%
%% Main text %%
%%
\section*{Main text}

We focus only on ``local'' mutations: mostly indel events (which may include local duplications),
also context-dependent substitutions

There is an extensive literature on nonlocal events such as rearrangements,
but these tend to defy phylogenetic reconstruction (due to the rapid proliferation of possible histories with even a few such events)

Two parts:
solving the master equation (Equations~\ref{MasterEquation}-\ref{MatrixExponential}),
extending from pairwise probability to a phylogenetic tree (Equations~\ref{Felsenstein}-\ref{Likelihood}).

\subsection*{Solving the master equation}

Exactly solvable models:
substitutions only,
TKF91
 - single-residue events only
 - linear birth-death model with immigration
 - Very appealing result: distribution of observed gap lengths is geometric
 - joint probability $P(\phi(0),\phi(t))$ expressible as a Pair HMM \cite{HolmesBruno2001}
 - conditional probability expressible as a weighted finite-state transducer \cite{Holmes2003,Westesson2012-zg,BouchardCote2013}

Problem with extending this to longer indel events is that a single gap may be consistent with
multiple indel ``trajectories'' (Figure)
Summing over all such trajectories to compute the probability of a particular size of gap being observed at a finite time
is a nontrivial problem

Models where the state is not a ``pure'' DNA or protein sequence,
but includes some extra artificial information,
such as ghost sites or hidden boundaries:
TKF92,
Metzler \cite{Metzler2003},
Rivas \cite{Rivas05},
Rivas \& Eddy \cite{RivasEddy2008,RivasEddy2015}

Considered as time-dependent probability distributions on pairwise alignments of
ancestor and descendant sequences,
these do not, in general, satisfy the Chapman-Kolmogorov forward equation
(\ref{ChapmanKolmogorov}).
That is, the probability of evolving from $i$ to $k$ comes out differently
depending on whether or not one conditions on an intermediate sequence $j$.
Clearly, something about this ``seems wrong'':
the failure to obey \eqref{ChapmanKolmogorov} illustrates the {\em ad hoc}
nature of these approaches.
Ezawa \cite{Ezawa2016b} describes the Chapman-Kolmogorov property as ``evolutionary consistency'';
in fact, \eqref{ChapmanKolmogorov} is the defining property
of any correct solution to a continuous-time Markov chain.
The above-mentioned approaches may be evolutionarily consistent if the state space
is allowed to include the extra information that is introduced to make the model tractable,
uch as the fragment boundaries in TKF92
or the ghost sites in \cite{RivasEddy2015}.

Miklos and Toroczkai insertion-only model(?) \cite{MiklosEtal2001}

Principled approximations that proceed systematically from Equations~\ref{MasterEquation}-\ref{MatrixExponential}
Lunter \& Hein's context-dependent substitution model (2004)
 - expansion of Taylor series for \eqref{MatrixExponential}
Miklos, Lunter and Holmes' ``long indel'' model (2004)
 - makes the intuitive claim that the existence of a conserved residue implies the alignment probability is factorable (because no indel has ever crossed that boundary)
 - uses a numerical sum over indel trajectories to approximate the probability distribution of observed gap lengths

More heuristic approaches which make less-principled single-event approximations.
Knudsen \& Miyamoto
Rivas
HandAlign
BaliPhy
ProtPal, IndelHistorian

Again, these approximations will not exactly satisfy \eqref{ChapmanKolmogorov}.
There will be some error in the probability, and in general the error will be greater
on longer branches, as the main assumption behind the approximation
(that there are no overlapping indels in the time interval, or that there is a finite
limit to the number of overlapping indels)
start to break down.
Since these are principled approximations, it is possible to study
how bad the error is, and this is the basis of the most exciting recent work in this area.


Simulation-based investigations
Rivas \& Eddy 2015
 - proof that an affine distribution for instantaneous indel events can't lead to an affine distribution for observed gap lengths, suggesting that HMM-like models may be fundamentally limited
 - simulations demonstrating this result
Ezawa 2016a,b
 - criticizes over-reliance on HMM-like models
 - puts the intuition of Miklos et al (2004) on a more formal footing by introducing an explicit notation (``local history set equivalence classes'') for evolutionary trajectories, and using it to prove that alignment is indeed factorable
 - investigates by simulation and analysis the relative contribution of multiple-event trajectories to gap length distributions


\subsection*{Extending from pairs to trees}

Most principled approach is MCMC - in principle can handle any indel length distribution.
Handel,
BaliPhy,
StatAlign,
HandAlign,
IndelHistorian.
In practise these all use HMMs

TKF91 model can be exactly extended to a tree \cite{Hein2001,LunterSongMiklosHein2003}

HMM-like nature of TKF91, and ubiquity of HMMs/DP in sequence analysis,
motivated numerous approaches based on Pair HMMs
\cite{ThorneEtal92,KnudsenMiyamoto2003,WangKeightleyJohnson2006,RivasEddy2008,RivasEddy2015}

Transducer approach is a general way of extending HMM-like pairwise alignment likelihoods to trees \cite{Westesson2012-zg,BouchardCote2013,IndelHistorian}

A finite-state transducer is a state machine that reads an input tape and writes to an output tape \cite{Mealy55}.
A probabilistically weighted finite-state transducer is the same, but its behavior is stochastic \cite{MohriPereiraRiley2000}
For the purposes of bioinformatics sequence analysis,
a transducer can be thought of as being just like a Pair HMM,
except where a Pair HMM's transition and emission probabilities
have been normalized so as to describe joint probabilities
like the entries of matrix $\matr{Q}(t)$ (\eqref{JointLikelihood}),
a transducer's probabilities are normalized so as to describe conditional probabilities
like the entries of matrix $\matr{M}(t)$ (\eqref{MatrixExponential}).
More specifically, if $i$ and $j$ are sequences, then the Forward score
for those two sequences using transducer $\trans{A}$
can be thought of as corresponding to the matrix entry $A_{ij}$.
Thus, the transducer is a compact encoding for a square matrix of countably infinite rank.

The utility of transducers is that for many purposes they can be manipulated just like matrices.
If $\trans{A}$ and $\trans{B}$ are transducers representing matrices $\matr{A}$ and $\matr{B}$,
then there is a well-defined operation called {\em transducer composition}
yielding a transducer $\trans{A}\trans{B}$ that represents the matrix $\matr{A}\matr{B}$.
There are other well-defined transducer operations corresponding to the Hadamard product
($\otimes$), the scalar product ($\cdot$) and the unit vector ($\Delta$),
so that Equations~\ref{Felsenstein}-\ref{Likelihood} can be interpreted directly
as transducer operations.

This has several benefits
Theoretical unification: \eqref{Felsenstein} is Sankoff's algorithm \cite{Sankoff85}
(excluding RNA structure; can be included by extending the transducer
framework to pushdown automata \cite{BradleyHolmes2009})
Phylogenetic alignment \cite{RedelingsSuchard2005,RedelingsSuchard2007}
Accurate parameter estimation \cite{Westesson2012-zg,Redelings2014}
Promising increases in accuracy of multiple sequence alignment \cite{IndelHistorian}

%%%%%%%%%%%%%%%%
%% Conclusions %%
%%
\section*{Conclusions}

Promise of using continuous-time Markov chains to parameterize evolutionary models of indels
has been partially realized by automata-theoretic approaches
based on transducers and HMMs.

Recent work by Rivas and Eddy \cite{RivasEddy2015}
and by Ezawa \cite{Ezawa2016a,Ezawa2016b,Ezawa2016bErratum}
may be interpreted as both good and bad news for automata-theoretic approaches.
It appears that closed-form probability distributions for observed gap lengths at finite times
under simple evolutionary models,
and in particular the geometric distributions that simple automata are good at modeling,
are still out of reach and indeed (for a simple case) proven impossible \cite{RivasEddy2015}.
Simulation results by both groups have demonstrated that geometric distributions are not
a good fit to the observed gap length distributions when the underlying indel model
has geometrically-distributed lengths for its instantaneous indel events.

However, the simulation results are mixed in the sense that the error does not appear
{\em too} bad.
% Ezawa results
Given the power of automata-theoretic approaches, the best way forward
(in the absence of a closed-form solution) may be to live with the error

Whatever approach is used, the benefits of modeling alignment as an aspect of
statistical phylogenetics are very strong



%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Backmatter begins here                   %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{backmatter}

\section*{Competing interests}
  The authors declare that they have no competing interests.

\section*{Author's contributions}
    Text for this section \ldots

\section*{Acknowledgements}
  Text for this section \ldots
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                  The Bibliography                       %%
%%                                                         %%
%%  Bmc_mathpys.bst  will be used to                       %%
%%  create a .BBL file for submission.                     %%
%%  After submission of the .TEX file,                     %%
%%  you will be prompted to submit your .BBL file.         %%
%%                                                         %%
%%                                                         %%
%%  Note that the displayed Bibliography will not          %%
%%  necessarily be rendered by Latex exactly as specified  %%
%%  in the online Instructions for Authors.                %%
%%                                                         %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% if your bibliography is in bibtex format, use those commands:
\bibliographystyle{bmc-mathphys} % Style BST file (bmc-mathphys, vancouver, spbasic).
\bibliography{references}      % Bibliography file (usually '*.bib' )
% for author-year bibliography (bmc-mathphys or spbasic)
% a) write to bib file (bmc-mathphys only)
% @settings{label, options="nameyear"}
% b) uncomment next line
%\nocite{label}

% or include bibliography directly:
% \begin{thebibliography}
% \bibitem{b1}
% \end{thebibliography}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                               %%
%% Figures                       %%
%%                               %%
%% NB: this is for captions and  %%
%% Titles. All graphics must be  %%
%% submitted separately and NOT  %%
%% included in the Tex document  %%
%%                               %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%
%% Do not use \listoffigures as most will included as separate files

\section*{Figures}
  \begin{figure}[h!]
  \caption{\csentence{Sample figure title.}
      A short description of the figure content
      should go here.}
      \end{figure}

\begin{figure}[h!]
  \caption{\csentence{Sample figure title.}
      Figure legend text.}
      \end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                               %%
%% Tables                        %%
%%                               %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%% Use of \listoftables is discouraged.
%%
\section*{Tables}
\begin{table}[h!]
\caption{Sample table title. This is where the description of the table should go.}
      \begin{tabular}{cccc}
        \hline
           & B1  &B2   & B3\\ \hline
        A1 & 0.1 & 0.2 & 0.3\\
        A2 & ... & ..  & .\\
        A3 & ..  & .   & .\\ \hline
      \end{tabular}
\end{table}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                               %%
%% Additional Files              %%
%%                               %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\section*{Additional Files}
  \subsection*{Additional file 1 --- Sample additional file title}
    Additional file descriptions text (including details of how to
    view the file, if it is in a non-standard format or the file extension).  This might
    refer to a multi-page table or a figure.

  \subsection*{Additional file 2 --- Sample additional file title}
    Additional file descriptions text.


\end{backmatter}
\end{document}
